pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c

```diff
@@ -0,0 +1,384 @@
+
+    int res;
+    struct netconn *tmp = NULL;
+
+    if (queue_len > USHRT_MAX) {
+        return -EFAULT;
+    }
+    if ((res = lwip_sock_create(&tmp, (struct _sock_tl_ep *)local, NULL, 0,
+                                flags, NETCONN_TCP)) < 0) {
+        return res;
+    }
+    assert(tmp != NULL); /* just in case lwIP is trolling */
+    mutex_init(&queue->mutex);
+    mutex_lock(&queue->mutex);
+    queue->conn = tmp;
+    queue->array = queue_array;
+    queue->len = queue_len;
+    queue->used = 0;
+    memset(queue->array, 0, sizeof(sock_tcp_t) * queue_len);
+    mutex_unlock(&queue->mutex);
+    switch (netconn_listen_with_backlog(queue->conn, queue->len)) {
+        case ERR_OK:
+            break;
+        case ERR_MEM:
+            return -ENOMEM;
+        case ERR_USE:
+            return -EADDRINUSE;
+        case ERR_VAL:
+            return -EINVAL;
+        default:
+            assert(false); /* should not happen since queue->conn is not closed
+                            * and we have a TCP conn */
+            break;
+    }
+    return 0;
+}
+
+void sock_tcp_disconnect(sock_tcp_t *sock)
+{
+    assert(sock != NULL);
+    mutex_lock(&sock->mutex);
+    if (sock->conn != NULL) {
+        netconn_close(sock->conn);
+        netconn_delete(sock->conn);
+        sock->conn = NULL;
+        /* if sock came from a sock_tcp_queue_t: since sock is a pointer in it's
+         * array it is also deleted from there, but we need to decrement the used
+         * counter */
+        if (sock->queue != NULL) {
+            assert(sock->queue->used > 0);
+            sock->queue->used--;
+            sock->queue = NULL;
+        }
+    }
+    mutex_unlock(&sock->mutex);
+    memset(&sock->mutex, 0, sizeof(mutex_t));
+}
+
+void sock_tcp_stop_listen(sock_tcp_queue_t *queue)
+{
+    assert(queue != NULL);
+    mutex_lock(&queue->mutex);
+    if (queue->conn != NULL) {
+        netconn_close(queue->conn);
+        netconn_delete(queue->conn);
+        queue->conn = NULL;
+        /* sever connections established through this queue */
+        for (unsigned i = 0; i < queue->len; i++) {
+            sock_tcp_disconnect(&queue->array[i]);
+        }
+        queue->array = NULL;
+        queue->len = 0;
+        queue->used = 0;
+    }
+    mutex_unlock(&queue->mutex);
+    memset(&queue->mutex, 0, sizeof(mutex_t));
+}
+
+int sock_tcp_get_local(sock_tcp_t *sock, sock_tcp_ep_t *ep)
+{
+    int res;
+    assert(sock != NULL);
+    mutex_lock(&sock->mutex);
+    if (sock->conn == NULL) {
+        return -EADDRNOTAVAIL;
+    }
+    else {
+        res = (lwip_sock_get_addr(sock->conn, (struct _sock_tl_ep *)ep,
+                                  1)) ? -EADDRNOTAVAIL : 0;
+    }
+    mutex_unlock(&sock->mutex);
+    return res;
+}
+
+int sock_tcp_get_remote(sock_tcp_t *sock, sock_tcp_ep_t *ep)
+{
+    int res;
+    assert(sock != NULL);
+    mutex_lock(&sock->mutex);
+    if (sock->conn == NULL) {
```

Comment:
Will do, but I prefer `sock->conn == NULL` over `!sock->conn` since it does not mix boolean operations with pointer checks